pkg/flow/module.go

```diff
@@ -2,19 +2,63 @@ package flow
 
 import (
 	"context"
+	"fmt"
 	"net/http"
 	"path"
 	"sync"
 
 	"github.com/gorilla/mux"
 	"github.com/grafana/agent/component"
 	"github.com/grafana/agent/pkg/cluster"
+	"github.com/grafana/agent/pkg/flow/internal/controller"
 	"github.com/grafana/agent/pkg/flow/logging"
+	"github.com/grafana/agent/pkg/flow/tracing"
 	"github.com/grafana/agent/web/api"
 	"github.com/prometheus/client_golang/prometheus"
-	"go.opentelemetry.io/otel/trace"
 )
 
+type moduleController struct {
+	mut sync.Mutex
+	o   *moduleControllerOptions
+	ids map[string]struct{}
+}
+
+var (
+	_ component.ModuleController = (*moduleController)(nil)
+)
+
+// newModuleController is the entrypoint into creating module instances.
+func newModuleController(o *moduleControllerOptions) component.ModuleController {
+	return &moduleController{
+		o:   o,
+		ids: map[string]struct{}{},
+	}
+}
+
+// NewModule creates a new, unstarted Module.
+func (m *moduleController) NewModule(id string, export component.ExportFunc) (component.Module, error) {
+	m.mut.Lock()
+	defer m.mut.Unlock()
+
+	if _, found := m.ids[id]; found {
+		return nil, fmt.Errorf("id %s already exists", id)
+	}
+	m.ids[id] = struct{}{}
```

Comment:
Its possible to move the ID into the ModuleController instead of the NewModule, which would ensure uniqueness at the cost of creating a ModuleController for each component. But is entirely doable.

Reply:
This does create a confusion if a component wants to create multiple modules since the ID must be unique. In our current cases we don't need that behavior and its best if the value is blank.

Reply:
> This does create a confusion if a component wants to create multiple modules since the ID must be unique. In our current cases we don't need that behavior and its best if the value is blank.

What do you mean? They already have to pass an ID to create multiple modules, and the ID must be unique to some scope; I just think it should only be locally unique to that component instance.

Reply:
PR for what it would look like https://github.com/grafana/agent/pull/4117/files.

Reply:
> PR for what it would look like https://github.com/grafana/agent/pull/4117/files.

That's fine with me. I'd have some feedback on the documentation for the interface to help make it more clear for users what the expectations are, but otherwise that looks good to help isolate components.

Reply:
Pulled it into this repo, would love any feedback on the documentation on the interface. Adjusted the docs a bit to indicate possible consequences

Reply:
Pulled it into this repo, would love any feedback on the documentation on the interface. Adjusted the docs a bit to indicate possible consequences